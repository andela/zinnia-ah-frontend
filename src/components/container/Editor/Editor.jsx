import React, { Component } from 'react';
import { EditorState, convertToRaw } from 'draft-js';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Editor } from 'react-draft-wysiwyg';
import draftToHtml from 'draftjs-to-html';
import { WithContext as ReactTags } from 'react-tag-input';
import { imageHandler } from '../../../utils/helpers.js';

//Styles
import '../../../../node_modules/react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import './Editor.scss';

//Reducer
import { createArticle } from '../../../store/modules/article';

//Components
import Button from '../../presentationals/Button/Button';
import Loader from '../../presentationals/Loader/Loader';
import TextArea from '../../presentationals/TextArea/TextArea';
import Navbar from '../../presentationals/Navbar/Navbar';

export class EditorContainer extends Component {
  state = {
    title: '',
    description: '',
    body: EditorState.createEmpty(),
    imageThumbnail: '',
    tags: [],
  };

  handleDelete = i => {
    const { tags } = this.state;
    this.setState({
      tags: tags.filter((tag, index) => index !== i),
    });
  };

  handleAddition = tag => {
    this.setState(state => ({ tags: [...state.tags, tag] }));
  };

  handleDrag = (tag, currPos, newPos) => {
    const tags = [...this.state.tags];
    const newTags = tags.slice();

    newTags.splice(currPos, 1);
    newTags.splice(newPos, 0, tag);

    // re-render
    this.setState({ tags: newTags });
  };

  onEditorStateChange = body => {
    this.setState({
      body,
    });
  };

  inputHandler = event => {
    this.setState({
      ...this.state,
      [event.target.name]: event.target.value,
    });
  };

  fileUploadHandler = async file => {
    const { imageThumbnail } = this.state;
    const data = await imageHandler(file);
    if (imageThumbnail === '') {
      this.setState({
        imageThumbnail: data.data.secure_url,
      });
    }
    return data;
  };

  submitArticle = event => {
    event.preventDefault();

    const getPreview = string => {
      return string.substring(0, 500);
    };

    const { title, body, imageThumbnail, tags } = this.state;
    const description = `${getPreview(
      JSON.stringify(body.getCurrentContent().getPlainText()),
    )}...`;

    const bodyCheck = `${getPreview(
      JSON.stringify(body.getCurrentContent().getPlainText()),
    )}`;

    let tagText = tags.map(tag => tag.text);

    const articleInput = {
      title,
      description,
      imageThumbnail,
      body: draftToHtml(convertToRaw(body.getCurrentContent())),
      tags: tagText,
    };

    if (bodyCheck === '""' && articleInput.title === '') {
      return false;
    }

    this.props.createArticle(articleInput);
  };

  render() {
    const {
      article: { isLoading },
    } = this.props;

    const { body, tags } = this.state;

    //Get unique keycodes
    const KeyCodes = {
      comma: 188,
      enter: 13,
    };

    const delimiters = [KeyCodes.comma, KeyCodes.enter];
    return (
      <div>
        {isLoading && <Loader text="please wait" size="large" />}
        <Navbar />
        <div className="editor-main">
          <TextArea
            autoFocus="autoFocus"
            rows="1"
            placeholder="Title"
            className="title-textarea"
            onKeyDown={this.enlargeTextarea}
            onChange={this.inputHandler}
            value={this.state.title}
            name="title"
          />
          <div className="editor-container">
            <Editor
              toolbarOnFocus
              editorState={body}
              placeholder="Write your post..."
              wrapperClassName="demo-wrapper"
              editorClassName="demo-editor"
              onEditorStateChange={this.onEditorStateChange}
              toolbar={{
                options: [
                  'inline',
                  'blockType',
                  'textAlign',
                  'colorPicker',
                  'link',
                  'emoji',
                  'image',
                  'history',
                ],
                inline: { inDropdown: true },
                list: { inDropdown: true },
                textAlign: { inDropdown: true },
                link: { inDropdown: true },
                history: { inDropdown: true },
                image: {
                  uploadCallback: this.fileUploadHandler,
                  placeHolder: 'Write your story...',
                  previewImage: true,
                  inputAccept:
                    'image/gif,image/jpeg,image/jpg,image/png,image/svg',
                  alignmentEnabled: false,
                  alt: { present: false, mandatory: false },
                  defaultSize: {
                    height: '500',
                    width: '1000',
                  },
                },
              }}
            />
            <div className="tags">
              <ReactTags
                tags={tags}
                handleDelete={this.handleDelete}
                handleAddition={this.handleAddition}
                handleDrag={this.handleDrag}
                delimiters={delimiters}
              />
            </div>
            <div className="d-flex justify-content-end">
              <div className="res-width-btn">
                <Button
                  type="submit"
                  value="POST ARTICLE"
                  className="btn-dark"
                  onClick={this.submitArticle}
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

EditorContainer.propTypes = {
  createArticle: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  article: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({ article: state.article });

export default connect(
  mapStateToProps,
  { createArticle },
)(EditorContainer);
