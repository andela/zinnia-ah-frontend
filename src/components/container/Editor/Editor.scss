.editor-container{
  border-color: #f00;
 border-style: solid;
}
