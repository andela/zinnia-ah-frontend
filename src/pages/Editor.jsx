import React, { Component } from 'react';
import Editor from '../components/container/Editor/Editor';

export class EditorContainer extends Component {
  render() {
    return (
      <div>
        <Editor />
      </div>
    );
  }
}

export default EditorContainer;
