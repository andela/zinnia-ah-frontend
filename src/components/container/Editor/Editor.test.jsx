import React from 'react';
import { shallow, mount } from 'enzyme';
import { EditorContainer } from './Editor';

jest.mock(
  '../../../../node_modules/draft-js/lib/generateRandomKey.js',
  () => () => '123',
);

const props = {
  article: {
    isLoading: true,
    successResponse: {
      status: 'success',
      message: '',
    },
    errorResponse: [],
  },
  createArticle: jest.fn(),
};

const wrapper = shallow(<EditorContainer {...props} />);
const MountWrapper = mount(<EditorContainer {...props} />);

describe('<EditorContainer />', () => {
  it('renders the component', () => {
    const input = MountWrapper.find('.ReactTags__tagInputField');
    input.instance().value = 'dc';
    input.simulate('change');
    MountWrapper.instance().handleDelete();
    MountWrapper.instance().handleDrag({ id: 'tat', text: 'tat' }, 3, 4);
    MountWrapper.instance().fileUploadHandler();
    expect(wrapper).toMatchSnapshot();
  });
});
