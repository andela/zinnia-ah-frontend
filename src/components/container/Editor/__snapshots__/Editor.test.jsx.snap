// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditorContainer /> renders the component 1`] = `
<div>
  <Loader
    size="large"
    text="please wait"
  />
  <div
    className="editor-main"
  >
    <TextArea
      autoFocus="autoFocus"
      className="title-textarea"
      name="title"
      onChange={[Function]}
      placeholder="Title"
      rows="1"
      value=""
    />
    <div
      className="editor-container"
    >
      <t
        customDecorators={Array []}
        editorClassName="demo-editor"
        editorState={
          EditorState {
            "_immutable": Immutable.Record {
              "allowUndo": true,
              "currentContent": Immutable.Record {
                "entityMap": Object {
                  "__add": [Function],
                  "__create": [Function],
                  "__get": [Function],
                  "__getLastCreatedEntityKey": [Function],
                  "__mergeData": [Function],
                  "__replaceData": [Function],
                  "add": [Function],
                  "create": [Function],
                  "get": [Function],
                  "getLastCreatedEntityKey": [Function],
                  "mergeData": [Function],
                  "replaceData": [Function],
                },
                "blockMap": Immutable.OrderedMap {
                  "123": Immutable.Record {
                    "key": "123",
                    "type": "unstyled",
                    "text": "",
                    "characterList": Immutable.List [],
                    "depth": 0,
                    "data": Immutable.Map {},
                  },
                },
                "selectionBefore": Immutable.Record {
                  "anchorKey": "123",
                  "anchorOffset": 0,
                  "focusKey": "123",
                  "focusOffset": 0,
                  "isBackward": false,
                  "hasFocus": false,
                },
                "selectionAfter": Immutable.Record {
                  "anchorKey": "123",
                  "anchorOffset": 0,
                  "focusKey": "123",
                  "focusOffset": 0,
                  "isBackward": false,
                  "hasFocus": false,
                },
              },
              "decorator": null,
              "directionMap": Immutable.OrderedMap {
                "123": "LTR",
              },
              "forceSelection": false,
              "inCompositionMode": false,
              "inlineStyleOverride": null,
              "lastChangeType": null,
              "nativelyRenderedContent": null,
              "redoStack": Immutable.Stack [],
              "selection": Immutable.Record {
                "anchorKey": "123",
                "anchorOffset": 0,
                "focusKey": "123",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
              "treeMap": Immutable.OrderedMap {
                "123": Immutable.List [
                  Immutable.Record {
                    "start": 0,
                    "end": 0,
                    "decoratorKey": null,
                    "leaves": Immutable.List [
                      Immutable.Record {
                        "start": 0,
                        "end": 0,
                      },
                    ],
                  },
                ],
              },
              "undoStack": Immutable.Stack [],
            },
          }
        }
        localization={
          Object {
            "locale": "en",
            "translations": Object {},
          }
        }
        onEditorStateChange={[Function]}
        placeholder="Write your post..."
        stripPastedStyles={false}
        toolbar={
          Object {
            "history": Object {
              "inDropdown": true,
            },
            "image": Object {
              "alignmentEnabled": false,
              "alt": Object {
                "mandatory": false,
                "present": false,
              },
              "defaultSize": Object {
                "height": "500",
                "width": "1000",
              },
              "inputAccept": "image/gif,image/jpeg,image/jpg,image/png,image/svg",
              "placeHolder": "Write your story...",
              "previewImage": true,
              "uploadCallback": [Function],
            },
            "inline": Object {
              "inDropdown": true,
            },
            "link": Object {
              "inDropdown": true,
            },
            "list": Object {
              "inDropdown": true,
            },
            "options": Array [
              "inline",
              "blockType",
              "textAlign",
              "colorPicker",
              "link",
              "emoji",
              "image",
              "history",
            ],
            "textAlign": Object {
              "inDropdown": true,
            },
          }
        }
        toolbarHidden={false}
        toolbarOnFocus={true}
        wrapperClassName="demo-wrapper"
      />
      <div
        className="tags"
      >
        <DragDropContext(ReactTags)
          delimiters={
            Array [
              188,
              13,
            ]
          }
          handleAddition={[Function]}
          handleDelete={[Function]}
          handleDrag={[Function]}
          tags={Array []}
        />
      </div>
      <div
        className="d-flex justify-content-end"
      >
        <div
          className="res-width-btn"
        >
          <Button
            className="btn-dark"
            onClick={[Function]}
            type="submit"
            value="POST ARTICLE"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
